library/src/main/java/org/xmtp/android/library/frames/FramesClient.kt

```diff
@@ -0,0 +1,92 @@
+
+    suspend fun signFrameAction(inputs: FrameActionInputs): FramePostPayload {
+        val opaqueConversationIdentifier = buildOpaqueIdentifier(inputs)
+        val frameUrl = inputs.frameUrl
+        val buttonIndex = inputs.buttonIndex
+        val inputText = inputs.inputText
+        val state = inputs.state
+        val now = System.currentTimeMillis() / 1000
+        val frameActionBuilder = FrameActionBody
+            .newBuilder()
+            .setFrameUrl(frameUrl)
+            .setButtonIndex(buttonIndex)
+            .setOpaqueConversationIdentifier(opaqueConversationIdentifier)
+            .setTimestamp(now)
+            .setUnixTimestamp(now.toInt())
+
+        if (inputText != null) {
+            frameActionBuilder.inputText = inputText
+        }
+        if (state != null) {
+            frameActionBuilder.state = state
+        }
+
+        val toSign = frameActionBuilder.build()
+        val signedAction = Base64.encodeToString(buildSignedFrameAction(toSign), Base64.NO_WRAP)
+
+        val untrustedData = FramePostUntrustedData(frameUrl, now, buttonIndex, inputText, state, xmtpClient.address, opaqueConversationIdentifier, now.toInt())
+        val trustedData = FramePostTrustedData(signedAction)
+
+        return FramePostPayload("xmtp@$PROTOCOL_VERSION", untrustedData, trustedData)
+    }
+
+    private suspend fun signDigest(digest: ByteArray): Signature {
+        val signedPrivateKey = xmtpClient.keys.identityKey
+        val privateKey = PrivateKeyBuilder.buildFromSignedPrivateKey(signedPrivateKey)
+        return PrivateKeyBuilder(privateKey).sign(digest)
+    }
+
+    private fun getPublicKeyBundle(): SignedPublicKeyBundle {
+        return xmtpClient.keys.getPublicKeyBundle()
+    }
+
+    private suspend fun buildSignedFrameAction(actionBodyInputs: FrameActionBody): ByteArray {
+        val digest = sha256(actionBodyInputs.toByteArray())
+        val signature = signDigest(digest)
+
+        val publicKeyBundle = getPublicKeyBundle()
+        val frameAction = FrameAction.newBuilder()
+            .setActionBody(actionBodyInputs.toByteString())
+            .setSignature(signature)
+            .setSignedPublicKeyBundle(publicKeyBundle).build()
+
+        return frameAction.toByteArray()
+    }
+
+    private fun buildOpaqueIdentifier(inputs: FrameActionInputs): String {
+        return when (inputs.conversationInputs) {
+            is ConversationActionInputs.Group -> {
+                val groupInputs = inputs.conversationInputs.inputs
+                val combined = groupInputs.groupId + groupInputs.groupSecret
+                val digest = sha256(combined)
+                Base64.encodeToString(digest, Base64.NO_WRAP)
+            }
+            is ConversationActionInputs.Dm -> {
+                val dmInputs = inputs.conversationInputs.inputs
+                val conversationTopic = dmInputs.conversationTopic ?: throw InvalidArgumentsError()
```

Comment:
Can we throw an XMTPException instead. This helps integrators catch exceptions being thrown from our SDK easier.
```suggestion
                val conversationTopic = dmInputs.conversationTopic ?: throw XMTPException("No conversation topic")
```